### PR TITLE
Feature: numeric ops errors

### LINKED ISSUE #64: Add error support to numeric literal ops
The (sparql) standard requires to raise "dynamic errors" under certain conditions in numeric operations which is currently not possible.
The problem is that the functions are marked noexcept and the return type does not allow for error value return.

There are two solutions:
1. remove the noexcept specification and just throw exceptions in the appropriate cases
    - trivial to implement
    - possibly bad performance
2. change the return type to be something like `std::expected<op_result_type, op_error>`
    - requires either: a. finding a library implementing something like it; or b. implementing our own until c++23 arrives

A decision needs to be made before merging some parts of the code for #45 (mainly related to one-side-bounded ap integer types, as there is no clear alternative to erroring on bound violation).

### PR DESCRIPTION
**Pull request by liss-h**
Fixes #64

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -37,6 +39,17 @@ if (NOT fmt_FOUND)
     message(DEBUG "Non CMake package for fmt found. fmt was retrieved via FetchContent.")
 endif ()
 
+if (NOT expected-lite_FOUND)
```
Comment (liss-h) [inline CMakeLists.txt:42]: Need some help here cannot figure out the correct cmake magic to make it work 😅
That's also why the (fetchContent) tests fail

Comment (bigerl) [reply CMakeLists.txt:42]: see https://github.com/rdf4cpp/rdf4cpp/pull/66/commits/9ca09f45cc3ae8d5d55863fe00a5c3497a07ca3c

You normally add only your own headers to `target_include_directories`. CMake takes care that  headers of dependencies are visible within the library and within the interface if required. 

If assets from dependencies are exposed via the public headers of our library, the dependencies must be in the `PUBLIC` section of `target_link_libraries`.

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp [hunk 2]
```diff
@@ -332,6 +332,15 @@ struct SelectOpRes<std::false_type, Fallback> {
     using type = Fallback;
 };
 
+template<typename T>
+[[nodiscard]] nonstd::expected<std::any, NumericOpError> map_expected(nonstd::expected<T, NumericOpError> const &e) noexcept {
+    if (e.has_value()) {
+        return *e;
+    } else {
+        return e.error();
```
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp]: Is this safe? Couldn't NumericOpError fall into `std::any`?

```suggestion
        return nonstd::make_unexpected(e.error());
```

**Review by bigerl — CHANGES_REQUESTED** [review]
Nice work. The code is very easy to follow and well understandable. 

Besides the one comment I have two more things that need to be taken care of before we merge:
- [x] double check that there is no portable implementation for `__builtin_x_overflow`
- [x] add tests that raise the errors

**Comment by liss-h** [comment]
> double check that there is no portable implementation for __builtin_x_overflow

I don't think there is, most compilers seem to have something with simmilar functionality but it's always slightly different.

Most people seem to just resort to a wall of macros that detects the compiler and then conditionally uses the correct function.
  - https://github.com/nemequ/portable-snippets/blob/master/safe-math/safe-math.h
  - https://github.com/dcleblanc/SafeInt

I guess we could use this kind of library for that but tbh it does not seem very robust to me. But I think this is probably a discussion for next meeting

**Comment by bigerl** [comment]
> Most people seem to just resort to a wall of macros that detects the compiler and then conditionally uses the correct function.
> 
>     * https://github.com/nemequ/portable-snippets/blob/master/safe-math/safe-math.h
> 
>     * https://github.com/dcleblanc/SafeInt

I tend towards safe-math as it makes a more mature impression to me. 

Edit: Boost also seems to have something: https://www.boost.org/doc/libs/1_80_0/libs/safe_numerics/doc/html/index.html

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp (hunk 2 above)
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp]: Yes, you are right. I have made that exact mistake multiple times now

**Comment by liss-h** [comment]
Since we are going to stick with builtins for overflow for now, there is nothing more to be done.

**Review by bigerl — APPROVED** [review]

**Pull request merged**